arirs/src/request_client/mod.rs

```diff
@@ -0,0 +1,120 @@
+        #[serde(flatten)]
+        inner: T,
+    }
+
+    #[derive(Debug, Getters, AsRef)]
+    pub struct RequestClient {
+        pub(crate) url: Url,
+        pub(crate) username: String,
+        pub(crate) password: String,
+        #[as_ref]
+        inner: reqwest::Client,
+    }
+
+    impl RequestClient {
+        pub(crate) fn new(url: Url, username: impl Into<String>, password: impl Into<String>) -> Self {
+            Self {
+                url,
+                username: username.into(),
+                password: password.into(),
+                inner: reqwest::Client::new(),
+            }
+        }
+
+        pub(crate) async fn authorized_get<T: Serialize, R: DeserializeOwned>(&self, path: impl AsRef<[&str]>, params: T) -> Result<R> {
+            let url = self.authorized_url(path, params)?;
+            let response = self.as_ref().get(url).send().await?.json().await?;
+            Ok(response)
+        }
+
+        pub(crate) async fn authorized_post<T: Serialize>(&self, path: impl AsRef<[&str]>, params: T) -> Result<()> {
+            let url = self.authorized_url(path, params)?;
+            self.as_ref().post(url).send().await?;
+            Ok(())
+        }
+
+        pub(crate) async fn authorized_post_json_response<T: Serialize, R: DeserializeOwned>(
+            &self,
+            path: impl AsRef<[&str]>,
+            params: T,
+        ) -> Result<R> {
+            let url = self.authorized_url(path, params)?;
+            let response = self.as_ref().post(url).send().await?.json().await?;
+            Ok(response)
+        }
+
+        pub(crate) async fn authorized_post_variables<T: Serialize, R: DeserializeOwned>(
+            &self,
+            path: impl AsRef<[&str]>,
+            params: T,
+            variables: &HashMap<&str, &str>,
+        ) -> Result<R> {
+            let url = self.authorized_url(path, params)?;
+            let response = self
+                .as_ref()
+                .post(url)
+                .json(&serde_json::json!({
+                    "variables": variables
+                }))
+                .send()
+                .await?
+                .json()
+                .await?;
+            Ok(response)
+        }
+
+        pub(crate) async fn authorized_delete<T: Serialize>(&self, path: impl AsRef<[&str]>, params: T) -> Result<()> {
+            let url = self.authorized_url(path, params)?;
+            self.as_ref().delete(url).send().await?;
+            Ok(())
+        }
+
+        fn authorized_url<'a, T: Serialize>(&self, path: impl AsRef<[&'a str]>, params: T) -> Result<Url> {
+            let mut url = self.url().join(&path.as_ref().join("/"))?;
+            self.set_authorized_query_params(&mut url, params);
+            Ok(url)
+        }
+
+        pub(crate) fn authorize_request<T>(&self, inner: T) -> AuthorizedRequest<T> {
+            AuthorizedRequest {
+                api_key: self.get_api_key(),
+                inner,
+            }
+        }
```

Comment:
question: why are all these named something with `authorize`? I may be wrong, but I don't know of any unauthenticated operations in ARI.

Reply:
fair :) I'll take a look at this once I've gained some more experience for working with the lib and ARI